RFC: Allow interfaces to implement other interfaces

mike-marcacci opened the pull request:
This is an implementation of the RFC described by spec PR https://github.com/graphql/graphql-spec/pull/373

This is a successor to PR #1218 which had accumulated too many merge conflicts to bother fixing.

mike-marcacci commented:
It looks like I can't request particular reviewers, so I'm just tagging @mjmahone and @michaelstaib in case you weren't notified already. The PR is incredibly simple but touches many pieces of code (especially tests).

mjmahone commented on src/type/schema.js:
```diff
@@ -234,7 +248,24 @@ export class GraphQLSchema {
+    return types;
+  }
+
+  getImplementations(
```
This should be a separate PR: this functionality is  useful even without interfaces implementing other interfaces, and should be considered on its own merits. Especially as you're not using it in this PR directly.

mike-marcacci replied:
Thanks so much for the excellent review! I just went to split these out, but their language and types _do_ assume that interfaces are potentially valid implementors. Are these assumptions appropriate to break out into an unrelated? I'm more than happy to do it, but I would be cautions of actually merging them without the rest, as the assumptions are only true in the context of this RFC.

mike-marcacci replied:
@mjmahone just to bump this and make sure I'm clear on the rationale for breaking this into a separate PR: is the thought that this is changing the public API of this type in a way that is useful even without the interface RFC? I am a bit reluctant to do this, because in a two-PR scenario I would introduce a non-breaking change in the first one (adding the methods, which only ever return concrete types simply because those are the only thing that can currently implement an interface); then introduce a potentially-breaking change in the second (these methods would now return both concrete types _and_ interfaces). Introducing these methods in the form written here makes sure there's only one, non-breaking change.

IvanGoncharov replied:
@mjmahone @mike-marcacci I don't think it's worth to extract this into separate PR since `getImplementations(interaceType)` is fully queal to `getPossibleTypes(interfaceType)` without this PR being merged.

mjmahone commented on src/type/schema.js:
```diff
@@ -254,6 +285,13 @@ export class GraphQLSchema {
     return Boolean(possibleTypeMap[abstractType.name][possibleType.name]);
   }
 
+  isImplementation(
```
Same here, though it's used once in this diff: we could either add an internal-only utility function for this for this PR, or land a PR with this included before this.

mike-marcacci replied:
(Same as above.) This function actually does call `getImplementations`; I broke them into two primarily to respect the pattern used by `getPossibleTypes`/`isPossibleType`.

mjmahone commented on src/language/__tests__/schema-parser-test.js:
```diff
@@ -480,6 +639,7 @@ describe('Schema Parser', () => {
           kind: 'InterfaceTypeDefinition',
           name: nameNode('Hello', { start: 10, end: 15 }),
           description: undefined,
+          interfaces: [],
```
Do we have an integration test (probably not under the `parser`, but in `buildASTSchema`) for an AST having `interfaces` as undefined?

I know we ran into problems in the past when hand-modeling the AST where, unless you explicitly set `directives: []`, you'd end up getting an undefined reference, even though the flow type allowed it.

mike-marcacci replied:
I just added an assertion for this case in dfc2da9d937987a7a23baf22ab203a8dea228087.

IvanGoncharov commented on src/type/schema.js:
```diff
@@ -123,7 +124,8 @@ export class GraphQLSchema {
   _typeMap: TypeMap;
-  _implementations: ObjMap<Array<GraphQLObjectType>>;
+  _implementations: ObjMap<Array<GraphQLObjectType | GraphQLInterfaceType>>;
+  _objectImplementations: ObjMap<Array<GraphQLObjectType>>;
```
@mike-marcacci It would be better to have a single object map:
```
_implementations: ObjectMap<{ objects: Array<GraphQLObjectType>, interfaces: Array<GraphQLInterfaceType> }>
```

mike-marcacci replied:
Hey @IvanGoncharov, I was implementing this just now, and I actually disagree here. Currently, `_implementations` contains a map of objects _and_ interfaces, while `_objectImplementations` only contains objects. Changing this so that interface implementations are stored separate from objects would make us have to recombine then in `getImplementations()` which would then return new arrays each time. This has the disadvantage of creating additional unnecessary objects (which probably doesn't matter much) but also breaks any object identity comparisons (`schema.getImplementations(myInterface) === schema.getImplementations(myInterface)`).

IvanGoncharov commented on src/__fixtures__/github-schema.json:
```diff
@@ -56052,7 +56052,7 @@
             {
               "name": "INTERFACE",
-              "description": "Indicates this type is an interface. `fields` and `possibleTypes` are valid fields.",
+              "description": "Indicates this type is an interface. `fields`, `interfaces`, and `possibleTypes` are valid fields.",
```
It's introspection snapshot of public GitHub API that we use for perf testing.
Please revert.

mike-marcacci replied:
Hmm, should this be reverted even though [the source line](https://github.com/mike-marcacci/graphql-js/blob/6f85d6b33c20c0de6556d153cd723dac4324472a/src/type/introspection.js#L401) has changed?

IvanGoncharov replied:
@mike-marcacci It's literally the result of introspection query against GitHub public API. Next time someone updates it this change will be lost.

mike-marcacci replied:
Ah, gotcha. Updated!

IvanGoncharov commented on src/__fixtures__/schema-kitchen-sink.graphql:
```diff
@@ -56,6 +56,12 @@ extend interface Bar {
 
 extend interface Bar @onInterface
 
+interface Baz implements Bar {
```
Since it's a kitchen sink it's better to implement couple of interfaces e.g. `implements Foo & Bar`.
Also please add an example of interface extension that adds a new interface

mike-marcacci replied:
Done and done.

IvanGoncharov commented on src/utilities/schemaPrinter.js:
```diff
@@ -194,9 +194,13 @@ function printObject(type: GraphQLObjectType, options): string {
 }
 
 function printInterface(type: GraphQLInterfaceType, options): string {
+  const interfaces = type.getInterfaces();
```
please extract it into `printImplementedInterfaces` and reuse for both objects and interfaces, similar to `printFields`

mike-marcacci replied:
👍

IvanGoncharov commented on src/utilities/buildClientSchema.js:
```diff
@@ -253,9 +253,17 @@ export function buildClientSchema(
+          inspect(interfaceIntrospection),
+      );
+    }
+
```
Please extract it into a separate function and reuse for building object type, similar to `buildFieldDefMap`

mike-marcacci replied:
What part here is reusable? Do you just want to consolidate the creation of this object?

```js
{
  name: interfaceIntrospection.name,
  description: interfaceIntrospection.description,
  interfaces: () => interfaceIntrospection.interfaces.map(getInterfaceType),
  fields: () => buildFieldDefMap(interfaceIntrospection),
}
```

If so, I'm not sure the extra redirection here is worth the DRY...

IvanGoncharov replied:
@mike-marcacci I'm referring to:
```
    if (!interfaceIntrospection.interfaces) {
      throw new Error(
        'Introspection result missing interfaces: ' +
          inspect(interfaceIntrospection),
      );
    }
    return interfaceIntrospection.interfaces.map(getInterfaceType)
```

mike-marcacci replied:
Ah got it; good call. Fixed.

IvanGoncharov commented on src/utilities/findBreakingChanges.js:
```diff
@@ -324,6 +324,30 @@ function findObjectTypeChanges(
+      type: BreakingChangeType.INTERFACE_REMOVED_FROM_OBJECT,
+      description: `${oldType.name} no longer implements interface ${oldInterface.name}.`,
+    });
+  }
```
Please extract into a separate function and reuse for the object type, similar to `findFieldChanges`

IvanGoncharov commented on src/utilities/findBreakingChanges.js:
```diff
@@ -190,7 +190,7 @@ function findTypeChanges(
       schemaChanges.push(...findObjectTypeChanges(oldType, newType));
     } else if (isInterfaceType(oldType) && isInterfaceType(newType)) {
-      schemaChanges.push(...findFieldChanges(oldType, newType));
+      schemaChanges.push(...findInterfaceTypeChanges(oldType, newType));
```
Don't think it's worth to have `findObjectTypeChanges` and `findInterfaceTypeChanges` as wrapper functions. Please change it to:
```
schemaChanges.push(
  ...findImplementedInterfacesChanges(oldType, newType),
  ...findFieldChanges(oldType, newType),
);
```
and do the same for object type.

mike-marcacci replied:
done

IvanGoncharov commented on src/utilities/schemaPrinter.js:
```diff
@@ -181,22 +181,27 @@ function printScalar(type: GraphQLScalarType, options): string {
   return (
     printDescription(options, type) +
-    `type ${type.name}${implementedInterfaces}` +
+    `type ${type.name}${printImplementedInterfaces(type)}` +
```
Please don't wrap it with `${}` use, please use simple concatenation same as `printFields` on a line bellow.

mike-marcacci replied:
done

IvanGoncharov commented on src/utilities/schemaPrinter.js:
```diff
@@ -181,22 +181,27 @@ function printScalar(type: GraphQLScalarType, options): string {
   return (
     printDescription(options, type) +
-    `interface ${type.name}` +
+    `interface ${type.name}${printImplementedInterfaces(type)}` +
```
same here

mike-marcacci replied:
done

mike-marcacci commented:
@IvanGoncharov I've added logic to prevent cyclic implementations. Because we already require flattening of transitively implemented interfaces, this ended up being trivial and didn't require any recursion. I believe this is in its final state.

I'm going to write up a blog post today, but this PR can probably be merged without it. Does anyone have a preference on how/where I publish the post? I can publish it myself on Medium after this is merged, or create a PR [to the GraphQL blog](https://github.com/graphql/graphql.github.io/tree/source/site/blog).

IvanGoncharov commented:
> Does anyone have a preference on how/where I publish the post? I can publish it myself on Medium after this is merged, or create a PR to the GraphQL blog

@mike-marcacci I think publishing it on personal account would be enough.

Also can you please update TS typings in `tstypes` folder?

> I believe this is in its final state.

Great 🎉 I'm currently in the middle of review:
![image](https://user-images.githubusercontent.com/8336157/64926038-71f59e80-d801-11e9-818a-1493fa8aff12.png)
But I will try to finish it tomorrow.

mike-marcacci commented:
Wonderful! Great catch on the TS types; I just added those.

IvanGoncharov commented on src/utilities/__tests__/buildASTSchema-test.js:
```diff
@@ -285,6 +285,12 @@ describe('Schema Builder', () => {
+    expect(
+      definition.kind === 'InterfaceTypeDefinition' && definition.interfaces,
+    ).to.deep.equal([], 'The interfaces property must be an empty array.');
+
```
@mike-marcacci Why do you need this check?
This should be checked in parser tests.

mike-marcacci replied:
Adding this test was requested by @mjmahone [in this comment](https://github.com/graphql/graphql-js/pull/2084#discussion_r313964490).

IvanGoncharov replied:
So Matt's comment was about testing if `buildSchema` supports both empty array and undefined:
> Do we have an integration test (probably not under the parser, but in buildASTSchema) for an AST having interfaces as undefined?

In its current form, you just duplicating parser tests since it's just checking `interfaces` value is an array and Matt suggested to modify it to `undefined` and test that result is the same.

At the moment we don't have any such test so think it should be a generic solution so I created #2203

mike-marcacci replied:
That's a great point; I assume I should just leave this as is and we can remove this test in a separate PR that fixes #2203?

IvanGoncharov commented:
@mike-marcacci Currently we are in feature freeze for `14.x.x` so this feature should be included in the upcoming `15.0.0`. I plan to release `15.0.0-alpha.1` in the next few days and hopefully include this feature.

I branch of `14.x.x` and merged `15.x.x` into `master`. Can you please rebase this PR on top of `master`?

mike-marcacci commented:
Thanks so much for the thorough review @IvanGoncharov! I just rebased this on top of master. 👍

I also added a `// prettier-ignore` to prevent the addition of a trailing comma in one of the typedefs.

IvanGoncharov commented:
@mike-marcacci Sorry for the delay I was traveling to a conference and speaking today.
I will 100% finish the review tomorrow.

mike-marcacci commented:
@IvanGoncharov no worries at all! You're an absolute hero for all your work here :)

IvanGoncharov commented on src/type/validate.js:
```diff
@@ -410,19 +429,53 @@ function validateObjectImplementsInterface(
+  implemented: GraphQLInterfaceType,
+): void {
+  const implementingInterfaces = implementing.getInterfaces();
+  implemented.getInterfaces().forEach(transitive => {
```
@mike-marcacci We don't use `forEach` can please use `for of` similar to other cycles in this file?

IvanGoncharov commented on src/execution/__tests__/union-interface-test.js:
```diff
@@ -140,7 +175,16 @@ describe('Execute: Union and intersection types', () => {
+          interfaces: [{ name: 'Life' }],
           possibleTypes: [{ name: 'Person' }, { name: 'Dog' }, { name: 'Cat' }],
           enumValues: null,
           inputFields: null,
```
Can you please also extend other tests bellow with named and inline fragments on intermidiate interfaces.

IvanGoncharov commented on src/language/__tests__/schema-parser-test.js:
```diff
@@ -226,13 +244,59 @@ describe('Schema Parser', () => {
+
+  it('Object extension do not include descriptions', () => {
+    expectSyntaxError(
+      `
```
Can you please change to 
```
expectSyntaxError(`
"Description"
...
`).to.deep.equal({
```

IvanGoncharov commented on src/language/__tests__/schema-parser-test.js:
```diff
@@ -247,6 +311,31 @@ describe('Schema Parser', () => {
 
+  it('Interface extension do not include descriptions', () => {
+    expectSyntaxError(
+      `
```
same here

IvanGoncharov commented on src/type/__tests__/schema-test.js:
```diff
@@ -183,24 +191,35 @@ describe('Type System: Schema', () => {
+      const SomeImplementingObject = new GraphQLObjectType({
+        name: 'SomeImplementingObject',
+        fields: {},
+        interfaces: () => [SomeInterface, SomeImplementingInterface],
```
No need to modify this test since you explicitly listing both interfaces here

mike-marcacci replied:
Hmm, unless I'm missing something, the only real changes here are to the names, which make them more symmetrical and consistent w/ the spec and the language used in the implementations.

mike-marcacci replied:
I updated all the names to match your suggestions... I'm not sure why GitHub's diffs aren't showing these changes. I'm going to force them to refresh.

IvanGoncharov replied:
@mike-marcacci Just to make sure we both on the same page: This test testing that all interfaces returned by `interfaces` Thunk included in the resulting schema. Since you all list all interfaces we can't test if schema traverse interfaces or not so add transitive interfaces doesn't test anything and just makes this test bigger.

mike-marcacci replied:
Aah, I see what you mean; point taken. I've reverted this file to match master.

IvanGoncharov commented on src/type/validate.js:
```diff
@@ -314,94 +317,110 @@ function validateFields(
+function validateInterfaces(
   context: SchemaValidationContext,
-  object: GraphQLObjectType,
+  implementing: GraphQLObjectType | GraphQLInterfaceType,
```
Can you please change the name to `type` similar to `validateFields`:
https://github.com/graphql/graphql-js/blob/5873fd26f60e88c473dea678a7723888f766658b/src/type/validate.js#L259-L262

IvanGoncharov commented on src/type/validate.js:
```diff
@@ -314,94 +317,110 @@ function validateFields(
+  implementing: GraphQLObjectType | GraphQLInterfaceType,
 ): void {
   const implementedTypeNames = Object.create(null);
-  for (const iface of object.getInterfaces()) {
```
It's unnecessary to rename `iface` to `implemented`.
It will remove unnecessary diff.

IvanGoncharov commented on src/type/validate.js:
```diff
@@ -314,94 +317,110 @@ function validateFields(
       continue;
     }
-    implementedTypeNames[iface.name] = true;
-    validateObjectImplementsInterface(context, object, iface);
```
It's better to use `validateTypeImplementsInterface`.

mike-marcacci replied:
I'm not 100% sure I'm following here (this removes those lines, and `validateTypeImplementsInterface` doesn't exist).

IvanGoncharov replied:
I mean if we rename `validateObjectImplementsInterface` to something `validateTypeImplementsInterface` is better follows existing naming schema than `validateImplementsAncestors`

mike-marcacci replied:
Ah, gotcha. As these are separate functions with the same signature, should these both be changed from:

```js
validateImplementsAncestors(context, type, iface);
validateImplementsInterface(context, type, iface);
```

to:
```js
validateTypeImplementsAncestors(context, type, iface);
validateTypeImplementsInterface(context, type, iface);
```

mike-marcacci replied:
(Previously `validateImplementsAncestors ` was only called from inside `validateImplementsInterface`, but this was changed to avoid an extra level of depth in the stack.)

mike-marcacci replied:
I've gone ahead and renamed these functions and variable names; hopefully these look better.

IvanGoncharov commented on src/type/validate.js:
```diff
@@ -314,94 +317,110 @@ function validateFields(
-  object: GraphQLObjectType,
-  iface: GraphQLInterfaceType,
+  implementing: GraphQLObjectType | GraphQLInterfaceType,
+  implemented: GraphQLInterfaceType,
```
Same here, I think `type` and `iface` is perfectly fine.
This also will minimize diff.
I agree that renaming stuff may be beneficial but it should be done in separate PR and for the entire file so we have consistency.

IvanGoncharov replied:
With all renaming going on it's extremely hard to spot actual changes :(

mike-marcacci replied:
Thi follows the language in the spec, which is very helpful when verifying correctness.

mike-marcacci replied:
With that in mind, would you still prefer this be in a separate PR?

mike-marcacci replied:
I've gone ahead and renamed these per your suggestion.

IvanGoncharov replied:
So there are two problems here:
1. Since now it validates both interface and objects but in your version, it matches only interface algorithm.
2. Maybe it's only me but `implementing` and `implemented` looks too similar when I'm scanning through the source code. Maybe it's because I'm not a native speaker or because of some mild dyslexia.

mike-marcacci replied:
1. Yes, originally I had two different functions: one for objects and one for interfaces. This allowed me to use language that matched the spec in both places. You're right, though, that since we combined them into one function, only one will have wording that matches the spec.
2. Yes, this is a bit hard to read, but in the context of interfaces implementing other interfaces, these were the most clear words I could think to use. I agree that while that might be a good choice in the spec (for the sake of correctness) it's much easier to follow here, and probably the right choice given point 1.

IvanGoncharov commented on src/type/validate.js:
```diff
@@ -410,19 +429,53 @@ function validateObjectImplementsInterface(
+    if (!implementingInterfaces.includes(transitive)) {
+      if (transitive === implementing) {
+        context.reportError(
+          `${isObjectType(implementing) ? 'Object' : 'Interface'} type ${
```
Please move this into
```
const typeKindStr = isObjectType(implementing) ? 'Object' : 'Interface';
```
so the rest of the string can be a single line

IvanGoncharov commented on src/type/validate.js:
```diff
@@ -410,19 +429,53 @@ function validateObjectImplementsInterface(
+  const implementingInterfaces = implementing.getInterfaces();
+  implemented.getInterfaces().forEach(transitive => {
+    if (!implementingInterfaces.includes(transitive)) {
+      if (transitive === implementing) {
```
This check is independent of the previous one so why did you nested it or am I missing something here?

mike-marcacci replied:
This is simply to switch which error message is generated (is the interface missing from the list of implemented interfaces due to a circular reference, or because it was never declared). I'm changing this to a ternary to make this logic more clear.

IvanGoncharov commented on src/type/validate.js:
```diff
@@ -314,94 +317,110 @@ function validateFields(
+  const implementedFieldMap = implemented.getFields();
+
+  // Assert each ancestor interface is explicitly implemented.
+  validateImplementsAncestors(context, implementing, implemented);
```
No need to make call stack deeper, please move it into `validateInterfaces` before `validateImplementsInterface` call.

IvanGoncharov commented on src/type/schema.js:
```diff
@@ -32,6 +32,7 @@ import {
   type GraphQLNamedType,
   type GraphQLAbstractType,
   type GraphQLObjectType,
```
You forget to synchronize this file with `schema.d.ts`

mike-marcacci replied:
Good catch!

mike-marcacci commented:
@IvanGoncharov thanks SO much for the great review. I pushed back on 2 of your comments RE name changes, since the language I used here mirrors the spec, and I found it much easier to follow when comparing the two. If you have a strong preference despite this, let me know and I'll modify the names here.

I did also need some clarification on one of the other comments.

Let me know about these and I'll get any other changes made ASAP!

IvanGoncharov commented on src/language/__tests__/schema-parser-test.js:
```diff
@@ -247,6 +309,29 @@ describe('Schema Parser', () => {
+      extend "Description" interface Hello {
+        world: String
+      }
+    `,
```
Can you please also reformat this one?

mike-marcacci commented:
@IvanGoncharov I've gone ahead and made all your suggested changes – we can debate the variable naming later. Let me know if you have any other requests, and I'll address them ASAP!

mike-marcacci commented:
Rebased on top of master again.

IvanGoncharov commented on tstypes/type/definition.d.ts:
```diff
@@ -387,6 +387,7 @@ export class GraphQLObjectType<
   toConfig(): GraphQLObjectTypeConfig<any, any> & {
     interfaces: GraphQLInterfaceType[];
     fields: GraphQLFieldConfigMap<any, any>;
+    extensions: Maybe<Readonly<Record<string, any>>>;
```
@mike-marcacci Thanks for fixing I went ahead so and fixed all similar stuff in #2202 sorry for rebases but it will keep cleaner diff in this PR.

mike-marcacci replied:
No worries – it's easy as long as I don't let it get too far out of date :)

IvanGoncharov commented on src/utilities/__tests__/findBreakingChanges-test.js:
```diff
@@ -572,6 +572,27 @@ describe('findBreakingChanges', () => {
+
+    expect(findBreakingChanges(oldSchema, newSchema)).to.deep.equal([
+      {
+        type: BreakingChangeType.INTERFACE_REMOVED_FROM_OBJECT,
```
It's a problem that enum value is called `INTERFACE_REMOVED_FROM_OBJECT` because it was removed from interface. Since we targeting `15.0.0` we can do breaking change so I propose to rename it to `IMPLEMENTED_INTERFACE_REMOVED` similar to `FIELD_REMOVED`:
https://github.com/graphql/graphql-js/blob/21de98ea4855df8c5146580611fda8ad76f52a2f/src/utilities/findBreakingChanges.js#L45

mike-marcacci replied:
Great point! I've renamed this to `IMPLEMENTED_INTERFACE_REMOVED`.

IvanGoncharov commented on src/utilities/__tests__/findBreakingChanges-test.js:
```diff
@@ -1000,6 +1021,30 @@ describe('findDangerousChanges', () => {
+
+    expect(findDangerousChanges(oldSchema, newSchema)).to.deep.equal([
+      {
+        type: DangerousChangeType.INTERFACE_ADDED_TO_OBJECT,
```
same here

mike-marcacci replied:
I've renamed this to `IMPLEMENTED_INTERFACE_ADDED` following the other types.

IvanGoncharov commented on src/type/schema.js:
```diff
@@ -250,6 +281,13 @@ export class GraphQLSchema {
+    interfaceType: GraphQLInterfaceType,
+    possibleType: GraphQLObjectType | GraphQLInterfaceType,
+  ): boolean {
+    return this.getImplementations(interfaceType).includes(possibleType);
```
Includes is not supported in IE11 so we can't use it until we drop support for it :(

IvanGoncharov replied:
Also please check other places for functions not supported in IE11

mike-marcacci replied:
I didn't realize that IE11 was a target for this library. I removed all instances of `Array.prototype.includes`, including one that wasn't introduced in this PR. I know that with TS you can specify the library available in the target... this is something we should make sure to set correctly during the move to TS to catch any of these cases.

mike-marcacci replied:
I'll keep my new changes from `.includes` to `.indexOf` for now, but reading through the babel configs, it looks like `.includes` should be fine, because it is rewritten with a polyfill when targeting IE11.

mike-marcacci replied:
Err... it looks like the other language features that would require an IE11-targeting app to include `babel-polyfill` are only used in the tests, so my claim here doesn't hold true; I reverted my change to the code not introduced by this PR, since it was in a test.

IvanGoncharov commented on src/type/validate.js:
```diff
@@ -410,19 +425,41 @@ function validateObjectImplementsInterface(
+  const implementingInterfaces = type.getInterfaces();
+  for (const transitive of iface.getInterfaces()) {
+    if (implementingInterfaces.indexOf(transitive) === -1) {
+      const typeKindStr = isObjectType(type) ? 'Object' : 'Interface';
```
We have other similar errors and they simply refer to type as `Type` I don't think this error should be special:
https://github.com/graphql/graphql-js/blob/21de98ea4855df8c5146580611fda8ad76f52a2f/src/type/validate.js#L268
https://github.com/graphql/graphql-js/blob/21de98ea4855df8c5146580611fda8ad76f52a2f/src/type/validate.js#L325
Fixed in https://github.com/mike-marcacci/graphql-js/pull/1

IvanGoncharov commented on src/type/validate.js:
```diff
@@ -410,19 +425,41 @@ function validateObjectImplementsInterface(
+): void {
+  const implementingInterfaces = type.getInterfaces();
+  for (const transitive of iface.getInterfaces()) {
+    if (implementingInterfaces.indexOf(transitive) === -1) {
```
This check doesn't for self-referencing interfaces:
```
interfaces Foo implements Foo {}
```
Fixed in https://github.com/mike-marcacci/graphql-js/pull/1

mike-marcacci replied:
Great catch!

IvanGoncharov commented on src/type/validate.js:
```diff
@@ -314,94 +317,106 @@ function validateFields(
     }
 
-    // Assert interface field type is satisfied by object field type, by being
+    // Assert interface field type is satisfied by implementing field type, by being
```
After the change, this comment doesn't make sense `is satisfied by implementing field type` is different from `is satisfied by object field type`.
Arguments here is the same as in this thread:
https://github.com/graphql/graphql-js/pull/2084#discussion_r328241103
So can we revert to the previous state and have `typeField` and `ifaceField`?
and thus minimize changes in this function.

IvanGoncharov commented:
Fixed type collection from interace's interfaces here:
https://github.com/mike-marcacci/graphql-js/pull/1/commits/887219c4dcb1bee87caa943e123e3e7aa43caf8b

IvanGoncharov commented on src/type/schema.js:
```diff
@@ -232,7 +246,24 @@ export class GraphQLSchema {
+    const types = this._objectImplementations[abstractType.name];
+    if (!types) {
+      return (this._objectImplementations[abstractType.name] = []);
+    }
```
What is the motivation for this change?
It's increasing the size `_objectImplementations` and also allocates memory during execution. If you feel like `|| []` is a problem (if so can you please explain why?) you can put an empty array inside the constructor to have predictable performance less chance to hit performance cliff due to V8 deoptimization.

mike-marcacci replied:
While this does add some trivial allocation of empty arrays, it prevents a new array from being created every time `getPossibleTypes` is called

mike-marcacci replied:
I'm certainly willing to move this to the constructor, but I'm not convinced that the resulting behavior would be an improvement. If this were a realtime language like rust or C, I would understand a preference for time-predictable calls here over total allocation, and then we could avoid memoization in general. However, given that JS is garbage collected and a memoization pattern was already employed here, I think this is the "minimal change" that actually provides correct behavior.

mike-marcacci replied:
Perhaps an alternative which would reduce allocations while still preservice object identity for comparison purposes would be to simply hoist the empty array out of the function. I'll implement that instead.

mike-marcacci replied:
I've changed my mind here RE the importance of object identity. While it's a goal that I appreciate, I don't want to force on a project where it's not a shared by everyone involved, as it would be too easy for a consuming app to rely on object identity for the sake of comparisons, then break on a minor/patch update if that isn't considered part of the API. I'm going to take your recommendation here and just use `|| []`.

IvanGoncharov replied:
I also changed my mind since interfaces and unions without subtypes should be pretty rare I moved this code into the constructor:
https://github.com/APIs-guru/graphql-js/blob/102a8d9eda29b6be52f37f4c17c857ae810895bb/src/type/schema.js#L335-L337

IvanGoncharov commented on src/utilities/buildClientSchema.js:
```diff
@@ -232,19 +232,28 @@ export function buildClientSchema(
+      | IntrospectionObjectType
+      | IntrospectionInterfaceType,
+  ) {
+    if (!implementingIntrospection.interfaces) {
```
We need to have legacy support for servers that would return `null` as `interfaces` value.
Fixed in https://github.com/mike-marcacci/graphql-js/pull/1/commits/07d8a2261f654768143a921507a1c96ce1cf645a

mike-marcacci commented:
@IvanGoncharov I just rebased on top of master to resolve a small merge conflict. I also squashed all the commits in this PR to make the rebase easier and so this can enter the history as a single atomic change. Just to make sure, is there anything you're waiting on from me?

IvanGoncharov commented:
@mike-marcacci Sorry for the delay. 
I reviewed almost all files and everything looks good.
But I want to try a couple of different approaches to `GraphQLSchema` changes.
I will try my hard to finish it tomorrow.

IvanGoncharov commented:
@mike-marcacci I created https://github.com/mike-marcacci/graphql-js/pull/2 with proposed changes:
1. `isPossibleType` & `isImplementation` merged into single `isSubType` function and deprecate `isPossibleType`.
2. `getImplementations` now returns object with `objects` and `interfaces` fields. That way we don't have two separate maps with overlapping arrays (one for objects and another for objects and interfaces).
Can you please review it?

It looks like it was the last major issue I spotted.

mike-marcacci commented:
I just merged your suggestions – thanks for making those as a PR :)

I agree that having a single map with wrapping objects is probably a better approach.
